index.bs

```diff
@@ -12403,158 +12294,157 @@ property is the String value "<code>forEach</code>".
-        *   the identifier "<code>entries</code>", and
-        *   the type "<code>method</code>".
-    1.  Let |interface| be the [=interface=]
-        on which the [=iterable declaration=] is declared.
-    1.  If |object| does not [=implement=] |interface|,
-        then [=ECMAScript/throw=] a {{ECMAScript/TypeError}}.
-    1.  Let |iterator| be a newly created [=default iterator object=]
-        for |interface| with |object| as its target and iterator kind "<code>value</code>".
-    1.  Return |iterator|.
-</div>
-
-<div algorithm="to invoke the values method of asynchronously iterable interfaces">
-
-    If the interface has an [=asynchronously iterable declaration=], then the method, when invoked,
-    must behave as follows:
-
-    1.  Let |object| be the result of calling [$ToObject$] on the <emu-val>this</emu-val> value.
-    1.  If |object| [=is a platform object=], then [=perform a security check=], passing:
-        *   the platform object |object|,
-        *   the identifier "<code>entries</code>", and
-        *   the type "<code>method</code>".
-    1.  Let |interface| be the [=interface=] on which the [=asynchronously iterable declaration=]
-        is declared.
-    1.  If |object| does not [=implement=] |interface|, then [=ECMAScript/throw=] a
-        {{ECMAScript/TypeError}}.
-    1.  Let |iterator| be a newly created [=default asynchronous iterator object=] for |interface|
-        with |object| as its [=default asynchronous iterator object/target=],
-        "<code>value</code>" as its [=default asynchronous iterator object/kind=], and
-        [=default asynchronous iterator object/is finished=] set to false.
-    1.  Run the [=asynchronous iterator initialization steps=] for |interface| with |object| and
-        |iterator|, if any such steps exist.
-    1.  Return |iterator|.
+<div algorithm>
+    To <dfn oldids="es-iterable-entries,es-iterable-keys,es-iterable-values">define the iterability</dfn>
+    of [=interface=] |definition| on |target|, given [=Realm=] |realm|, run the following steps:
+
+    1.  If |definition| has an [=indexed property getter=], then:
+        1.  Perform [=!=] [$CreateMethodProperty$](|target|, {{@@iterator}},
+            {{%ArrayProto_values%}}).
+    1.  If |definition| has a [=value iterator=], then:
+        1.  Assert: |definition| has an [=indexed property getter=].
+        1.  Perform [=!=] [$CreateDataProperty$](|target|, "<code>entries</code>",
+            {{%ArrayProto_values%}}).
+        1.  Perform [=!=] [$CreateDataProperty$](|target|, "<code>keys</code>",
+            {{%ArrayProto_keys%}}).
+        1.  Perform [=!=] [$CreateDataProperty$](|target|, "<code>values</code>",
+            {{%ArrayProto_values%}}).
+        1.  Perform [=!=] [$CreateDataProperty$](|target|, "<code>forEach</code>",
+            {{%ArrayProto_forEach%}}).
+    1.  If |definition| has a [=pair iterator=], then:
+        1.  Define the {{@@iterator}} and <code>entries</code> methods:
```

Comment:
Is this "define the … methods" thing defined anywhere? Other places are using CreateBuiltinFunction which seems to be what this should do too.

Reply:
They are just headers grouping a bunch of steps together. The CreateBuiltinFunction calls are nested under these.

Reply:
Ah got it. This makes me think that there's value in the readability and navigability. The whole algorithm creates seven different functions, without anchors for each. In ES, this would correspond to seven sections; while here, it's one section. I think the optimal value is somewhere in between. (Putting everything into a single algorithm/section would be even worse for maplike/setlike: each one of those declarations creates 6–7 methods.)

Maybe something more similar to the original section layout (except the `entries`, `keys`, and `values` sections are renamed to "`entries` for pair iterators" and so on), and in the root section do something like

1. Let _steps_ be the algorithm steps defined in §&nbsp;3.7.1 `entries` for pair iterators.
2. Let _F_ be ! CreateBuiltinFunction(_steps_, « », _realm_).

Reply:
For async iterable declarations, I feel like they should be in a separate heading from §&nbsp;3.6.9 Iterable declarations anyway. They are not that similar.

Reply:
This turns out to be difficult because the steps close over _definition_. We can smuggle it in as an internal slot on the function but that's a lot of effort.

Ultimately I'm not sure giving these sections is that helpful:

- The steps are fairly repetitive, with the non-forEach ones varying only in key/value/key+value.
- You still have 4-5 lines per method of CreateBuiltinFunction, SetFunctionName, SetFunctionLength, and CreateDataProperty/CreateMethodProperty. That is, if we move the steps out, then we just have 17 steps next to each other repeating calls to those five abstract ops. Which is harder to read, IMO, than the current version.

Reply:
We could split up the iterable and async iterable sections, though; that seems reasonable. I guess then we'd have "define the iteration methods" and "define the asynchronous iteration methods".

Reply:
I think this is good now that a big chunk has been moved out of this algorithm.